### PR TITLE
EPMRPP-82351 || Add migration of merged launches

### PR DESCRIPTION
**Pull request by IvanKustau**

### CODE REVIEW

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java [hunk 1]
```diff
@@ -24,65 +24,120 @@
...
-		return personJsonObject;
-	}
-
-	private HttpEntity<String> getStringHttpEntity(String body) {
-		HttpHeaders headers = new HttpHeaders();
-		headers.setContentType(MediaType.APPLICATION_JSON);
-
-		return new HttpEntity<>(body, headers);
-	}
+  private final static String SEARCH_QUERY_JSON =
```
Comment (rkukharenka) [inline src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: line 27,30 - static modifier out of order (private **static** final)

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java [hunk 2]
```diff
@@ -24,65 +24,120 @@
...
+
+  public Optional<LogMessage> getFirstLogFromElasticSearch() {
+
+    SearchResponse searchResponse =
+        restTemplate.postForObject(host + "/_search", getStringHttpEntity(SEARCH_QUERY_JSON),
+            SearchResponse.class
+        );
+
+    if (searchResponse != null && searchResponse.getHits() != null
+        && searchResponse.getHits().getHits() != null && !searchResponse.getHits().getHits()
```
Comment (rkukharenka) [inline src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: you can replace  `searchResponse.getHits().getHits() != null && !searchResponse.getHits().getHits().isEmpty()` to `!CollectionUtils.isEmpty(searchResponse.getHits().getHits())`

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java [hunk 3]
```diff
@@ -24,65 +24,120 @@
...
+
+  public Optional<LogMessage> getLogFromLaunch(Long launchId, Long projectId) {
+    String indexName = "logs-reportportal-" + projectId;
+    String query = SEARCH_BY_LAUNCH_ID_JSON.replace("?", launchId.toString());
+    SearchResponse searchResponse =
+        restTemplate.postForObject(host + "/" + indexName + "/_search?size=1",
+            getStringHttpEntity(query), SearchResponse.class
+        );
+
+    if (searchResponse != null && !CollectionUtils.isEmpty(searchResponse.getHits().getHits())) {
```
Comment (rkukharenka) [inline src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: At this line searchResponse.getHits()**.getHits()** we can have potential NPE if first call getHits will return null. This case you predicted at the line 70

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java [hunk 4]
```diff
@@ -24,65 +24,120 @@
...
+    personJsonObject.put("id", logMessage.getId());
+    personJsonObject.put("message", logMessage.getLogMessage());
+    personJsonObject.put("itemId", logMessage.getItemId());
+    personJsonObject.put("@timestamp", logMessage.getLogTime());
+    personJsonObject.put("launchId", logMessage.getLaunchId());
+
+    return personJsonObject;
+  }
+
+  private JSONObject getUpdateLaunchIdJson(List<Long> testItemIds, Long newLaunchId) {
```
Comment (rkukharenka) [inline src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java:115]: seems you can rewrite method (lines 120-133)

```
    return new JSONObject()
        .put("script", "ctx._source.launchId = " + newLaunchId)
        .put("query", new JSONObject()
            .put("terms", new JSONObject()
                .put("itemId", new JSONArray(testItemIds))));
```

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java [hunk 5]
```diff
@@ -24,65 +24,120 @@
...
+
+  private JSONObject convertToJson(LogMessage logMessage) {
+    JSONObject personJsonObject = new JSONObject();
+    personJsonObject.put("id", logMessage.getId());
+    personJsonObject.put("message", logMessage.getLogMessage());
+    personJsonObject.put("itemId", logMessage.getItemId());
+    personJsonObject.put("@timestamp", logMessage.getLogTime());
+    personJsonObject.put("launchId", logMessage.getLaunchId());
+
+    return personJsonObject;
```
Comment (rkukharenka) [inline src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: More atractive variant:
```
    return new JSONObject()
        .put("id", logMessage.getId())
        .put("message", logMessage.getLogMessage())
        .put("itemId", logMessage.getItemId())
        .put("@timestamp", logMessage.getLogTime())
        .put("launchId", logMessage.getLaunchId());
```

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java [hunk 6]
```diff
@@ -24,65 +24,120 @@
...
+      return Optional.of(searchResponse.getHits().getHits().get(0).getSource());
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  public void updateLogsLaunchIdByTestItemId(List<Long> testItemIds, Long launchId,
+      Long projectId) {
+    String indexName = "logs-reportportal-" + projectId;
+    if (CollectionUtils.isEmpty(testItemIds)) {
```
Comment (rkukharenka) [inline src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java:96]: it makes sense to move up the condition higher on one line

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java (hunk 1 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: Updated

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java (hunk 2 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: Replaced

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java (hunk 3 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: Thank you for highlighting, added null check

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java (hunk 4 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java:115]: Rewritten

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java (hunk 5 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java]: Rewritten

File: src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java (hunk 6 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/elastic/SimpleElasticSearchClient.java:96]: Moved condition higher

**Review by rkukharenka — APPROVED** [review]

File: src/main/java/com/epam/reportportal/service/impl/ElasticMigrationServiceImpl.java [hunk 7]
```diff
@@ -13,176 +23,238 @@
...
+        closestLogId = namedParameterJdbcTemplate.queryForObject(SELECT_LOG_ID_CLOSEST_TO_TIME,
+            Map.of("time", startDateTime), Long.class
+        );
+      } catch (EmptyResultDataAccessException e) {
+        closestLogId = (long) Integer.MAX_VALUE;
+      }
+      LOGGER.info("Closest to start time log id : {}", closestLogId);
+      compareIdsAndMigrate(databaseFirstLogId, closestLogId);
+    } else {
+      LOGGER.info("");
```
Comment (APiankouski) [inline src/main/java/com/epam/reportportal/service/impl/ElasticMigrationServiceImpl.java]: Empty log

**Review by APiankouski — APPROVED** [review]

File: src/main/java/com/epam/reportportal/service/impl/ElasticMigrationServiceImpl.java (hunk 7 above)
Comment (IvanKustau) [reply src/main/java/com/epam/reportportal/service/impl/ElasticMigrationServiceImpl.java]: Removed

**Review by APiankouski — APPROVED** [review]

**Pull request merged**